### ISSUE TITLE
new metric that optimizes for speed

### ISSUE THREAD

**Issue by breznak** [numenta/nupic#1711]
Reference: #1363

**Comment by rhyolight** [comment]
@breznak I'd like to see more detail on this ticket description.

**Comment by rhyolight** [comment]
@breznak Thanks! Much better!

**Comment by chetan51** [comment]
@breznak Here is a proposed plan for adding additional metrics to swarm over (in this case, speed) and a notion of secondary metrics, in a minimally invasive way:

1. Add an optional `result` (ModelResult) field to MetricsIface.addInstance()
2. Add an optional `duration` field to ModelResult
3. CLAModel.run() sets the `duration` field on ModelResult
4. MetricsManager.update(result) passes the ModelResult in when calling Metric.addInstance()
5. SpeedMetric.addInstance() can use the `duration` field in the ModelResult instance to compute average duration as its metric value
6. Modify ModelRunner to accept in __init__ a list of metric key patterns (instead of a single `_optimizeKeyPattern`), and store these as a list of metrics to optimize for
7. Modify ModelRunner.__checkIfBestCompletedModel() to use the list of metrics to optimize for, and check them in order, using later metrics only when earlier metrics result in a tie
8. Repeat step 2 and step 5 when adding other types of data for metrics to use

**Comment by breznak** [comment]
@chetan51 a couple of comments..
- points 6.,7. : we shouldn't need this, I've created the `MultiMetric` just for that. Setting the secondary metric's weight to something tine makes it work just like the tie-breaker.
- point 5., `SpeedMetric` should not need this access, I've work-arounded it quite nice in https://github.com/numenta/nupic/pull/1726
- I'm not clear about the idea of "enable support for not-upstream (custom) metrics" - I will need to keep a branch with the metric anyway, so what's the matter of just patching the existing `metrics.py` and adding it there?
- This still does not fix some other metrics - namely `AnomalyMetric` that needs to get anomalyScores somehow.
 - I like the idea in points 1-4.! but instead of duration, I'd suggest adding a reference to itself (Model) to ModelResult, something like`ModelResult.parentModel` - this allows any possible future hacks in metrics.

**Comment by chetan51** [comment]
@breznak 

> - points 6.,7. : we shouldn't need this, I've created the `MultiMetric` just for that. Setting the secondary metric's weight to something tine makes it work just like the tie-breaker.

Okay, if this works already, then we can ignore points 6 and 7.

> - point 5., `SpeedMetric` should not need this access, I've work-arounded it quite nice in https://github.com/numenta/nupic/pull/1726

I don't think this is a good workaround, because it's not guaranteed from the Metric API that you'll compute the correct duration that way. Only the CLAModel's `run` function knows accurately how long it took to run one iteration, so it should be the one to store that information in the ModelResult.

> - I'm not clear about the idea of "enable support for not-upstream (custom) metrics" - I will need to keep a branch with the metric anyway, so what's the matter of just patching the existing `metrics.py` and adding it there?

I'm sorry, I don't understand your question.

> - This still does not fix some other metrics - namely `AnomalyMetric` that needs to get anomalyScores somehow.

If we need anomaly scores in the metrics, we would add them to ModelResult explicitly just like we would add `duration`.

> - I like the idea in points 1-4.! but instead of duration, I'd suggest adding a reference to itself (Model) to ModelResult, something like`ModelResult.parentModel` - this allows any possible future hacks in metrics.

I think it's better to have the Model explicitly add any data we need into the ModelResult it creates, rather than having a reference to the Model in the ModelResult. Otherwise we will find ourselves storing data like duration and anomaly scores in the model instance, and then accessing them hackily in the ModelResult, creating a messy 2-way dependency between the classes.

In other words, to keep the design clean, ModelResult shouldn't depend on Model.

**Comment by breznak** [comment]
> If we need anomaly scores in the metrics, we would add them to ModelResult explicitly just like we would add duration.

This is a very clean solution from the view-point of the Metrics. I'm not sure if @scottpurdy wasn't against cluttering the ModelResult with additional info at the beginning of these PRs (?) - the problem is Model is not aware of instantionized metrics - so should it append always `duration`? (`anomalyScore` can be added only for `TemporalAnomaly` model where it makes sense)

**Comment by chetan51** [comment]
> the problem is Model is not aware of instantionized metrics - so should it append always `duration`

Yes, I think it should always add duration, since that is relevant information for a ModelResult.

**Comment by breznak** [comment]
>> the problem is Model is not aware of instantionized metrics - so should
it append always duration

> Yes, I think it should always add duration, since that is relevant
information for a ModelResult.


great, I'll rework both metrics accordingly. Then the "expose model .." PR
would be obsolete (there are still some cleanups that might be worth
picking)
